### PR TITLE
feat(developer): kmc-package support remote fonts and files 🐻

### PR DESCRIPTION
**Pull request by mcdurdin**
Note: epic/shared-fonts is brought to you by the Share Bear 🐻

The concept here is that the 'Name' property for a file can now be a remote stable reference to a file stored on GitHub, rather than a local file. Additionally, we add a 'Source' property that may change over time, so we can be notified of changes to files.

## File.Source

The `Name` element may be either a local filename, or a GitHub raw permanent
URL, conforming to the pattern:

```
<Name>https://github.com/{owner}/{repo}/raw/{commit}/[path/]{filename}</Name>
```

We will add a `Source` element to .kps `Package.Files.File`. We would support
several sources:
* local - local file system
* flo - fonts.languagetechnology.org
* github
* noto - in the future

## local: `<Source>` element is omitted

This is the same as previous versions. `<Name>` references a file on the local
filesystem. Note that `<Source>` is optional for other sources as well; the
absence of the element does not mean local -- that is determined by pattern
match.

```xml
<File>
  <Name>[relative-or-absolute-path/]{filename}</Name>
</File>
```

#### Example

```xml
<File>
  <Name>Alkalami-Regular.ttf</Name>
</File>
```

```xml
<File>
  <Name>../../shared/fonts/Alkalami-Regular.ttf</Name>
</File>
```

### fonts.languagetechnology.org `flo:<id>`

References a font by ID from fonts.languagetechnology.org. The resource must be
resolved to a permanent URL. If the `<Name>` field is omitted, the compiler will
return an error, reporting the URL it discovers, so the author should be able to
paste that URL in.

If the URL in the `<Name>` field differs from the resolved URL given by FLO,
then a hint will be issued by the compiler, allowing the author to easily update
to a new version of the resource at their convenience. It is anticipated that
this may be scripted on the keyboards repository in the future.

```xml
<File>
  <Name>https://github.com/{owner}/{repo}/raw/{commit}/[path/]{filename}</Name>
  <Source>flo:{id}</Source>
</File>
```

#### Example

```xml
<File>
  <Name>https://github.com/silnrsi/fonts/raw/b88c7af5d16681bd137156929ff8baec82526560/fonts/sil/alkalami/Alkalami-Regular.ttf</Name>
  <Source>flo:alkalami</Source>
</File>
```

### GitHub

A reference to a resource located on GitHub. The resource must be resolved to a
permanent URL. If the `<Name>` field is omitted, the compiler will return an
error, reporting the URL it discovers, so the author should be able to paste
that URL in.

If the URL in the `<Name>` field differs from the resolved URL from the
`<Source>` element, then a hint will be issued by the compiler, allowing the
author to easily update to a new version of the resource at their convenience.
It is anticipated that this may be scripted on the keyboards repository in the
future.

```xml
<File>
  <Name>https://github.com/{owner}/{repo}/raw/{commit}/[path/]{filename}</Name>
  <Source>https://github.com/{owner}/{repo}/(raw|blob)/([refs/heads/]{branch}|[refs/tags/]{tag}|{commit})/[path/]{filename}</Source>
</File>
```

#### Examples

A GitHub raw reference, from a commit, resolves identically.

```xml
<File>
  <Name>https://github.com/silnrsi/fonts/raw/b88c7af5d16681bd137156929ff8baec82526560/fonts/sil/alkalami/Alkalami-Regular.ttf</Name>
  <Source>https://github.com/silnrsi/fonts/raw/b88c7af5d16681bd137156929ff8baec82526560/fonts/sil/alkalami/Alkalami-Regular.ttf</Source>
</File>
```

A GitHub raw reference, from a branch, with refs/heads/ excluded:

```xml
<File>
  <Name>https://github.com/silnrsi/fonts/raw/b88c7af5d16681bd137156929ff8baec82526560/fonts/sil/alkalami/Alkalami-Regular.ttf</Name>
  <Source>https://github.com/silnrsi/fonts/raw/main/fonts/sil/alkalami/Alkalami-Regular.ttf</Source>
</File>
```

A GitHub raw reference, from a branch, with refs/heads/ included:

```xml
<File>
  <Name>https://github.com/silnrsi/fonts/raw/b88c7af5d16681bd137156929ff8baec82526560/fonts/sil/alkalami/Alkalami-Regular.ttf</Name>
  <Source>https://github.com/silnrsi/fonts/raw/refs/heads/main/fonts/sil/alkalami/Alkalami-Regular.ttf</Source>
</File>
```

A GitHub blob reference, as copied from a URL in GitHub:

```xml
<File>
  <Name>https://github.com/silnrsi/fonts/raw/b88c7af5d16681bd137156929ff8baec82526560/fonts/sil/alkalami/Alkalami-Regular.ttf</Name>
  <Source>https://github.com/silnrsi/fonts/blob/main/fonts/sil/alkalami/Alkalami-Regular.ttf</Source>
</File>
```

### noto

In a future version.

## Additional file format notes

At the same time `CopyLocation`, `Description`, and `FileType` elements should
be deprecated and totally ignored by the compiler. They are busydata.

Future sources could be considered, e.g. noto. We don't want to allow arbitrary URLs, both for stability and for security reasons.

---

This change is entirely compiler-side, so we don't need to make any changes to apps, and so packages will be backwardly compatible. 

# Things to do

--> moved to #11236

Replaces: #12336
Fixes: #11236

@keymanapp-test-bot skip

### CODE REVIEW

**Comment by mcdurdin** [comment]
Capturing discussion on font versioning. We need to think this through with FLO, caching of fonts, versioning updates.


Marc Durdin
  30 minutes ago
in [fonts.languagetechnology.org](http://fonts.languagetechnology.org/), where is the version data sourced from in the .ttf? I see vast majority of fonts have version #.### format, but three fonts have semantic version patterns:  Dukor dukor (1.0.7), suranna (1.0.5) and wakor (4.0.7).
27 replies


Marc Durdin
  [29 minutes ago](https://sil-lt.slack.com/archives/CJ6C33JS0/p1731488242157679?thread_ts=1731488216.311579&cid=CJ6C33JS0)
We want to see if we need to re-download a font based on the version metadata provided, and also verify font version


Victor Gaultney
  20 minutes ago
Here’s a summary: https://silnrsi.github.io/FDBP/en-US/Versioning.html


Marc Durdin
  [20 minutes ago](https://sil-lt.slack.com/archives/CJ6C33JS0/p1731488776209109?thread_ts=1731488216.311579&cid=CJ6C33JS0)
So looks like you are using the name table
4:07
Thanks!


Victor Gaultney
  19 minutes ago
Version info is in both head and name
4:08
decimal and string
4:09
The three fonts you mention are not our and are out of step with the common #.### industry pattern


Marc Durdin
  [17 minutes ago](https://sil-lt.slack.com/archives/CJ6C33JS0/p1731489011883119?thread_ts=1731488216.311579&cid=CJ6C33JS0)
Yeah
[4:10](https://sil-lt.slack.com/archives/CJ6C33JS0/p1731489037637119?thread_ts=1731488216.311579&cid=CJ6C33JS0)
I just downloaded wakor, and its name record says Version 4.007 so something seems inconsistent there


Victor Gaultney
  [15 minutes ago](https://sil-lt.slack.com/archives/CJ6C33JS0/p1731489093952349?thread_ts=1731488216.311579&cid=CJ6C33JS0)
You can try to map the semantic-style to the industry standard using the M.mpp format
[4:11](https://sil-lt.slack.com/archives/CJ6C33JS0/p1731489114243179?thread_ts=1731488216.311579&cid=CJ6C33JS0)
So 4.0.7 is really 4.007
[4:12](https://sil-lt.slack.com/archives/CJ6C33JS0/p1731489158721159?thread_ts=1731488216.311579&cid=CJ6C33JS0)
By definition the  head must be #.#


Marc Durdin
  [13 minutes ago](https://sil-lt.slack.com/archives/CJ6C33JS0/p1731489209303039?thread_ts=1731488216.311579&cid=CJ6C33JS0)
But that doesn't explain where 4.0.7 is coming from in FLO data -- given the .ttf itself has 4.007 (both tables), shouldn't FLO be using that?
[4:14](https://sil-lt.slack.com/archives/CJ6C33JS0/p1731489258152699?thread_ts=1731488216.311579&cid=CJ6C33JS0)
Suranna-Regular.ttf has Version 1.0.5; .... and head.fontRevision 0x10000 which really doesn't help.


Victor Gaultney
  [12 minutes ago](https://sil-lt.slack.com/archives/CJ6C33JS0/p1731489268625869?thread_ts=1731488216.311579&cid=CJ6C33JS0)
That comes from the documentation from the source, not the ttf


Marc Durdin
  [12 minutes ago](https://sil-lt.slack.com/archives/CJ6C33JS0/p1731489284249799?thread_ts=1731488216.311579&cid=CJ6C33JS0)
I see. So we cannot reliably automatically check then. That's a shame
[4:16](https://sil-lt.slack.com/archives/CJ6C33JS0/p1731489365075489?thread_ts=1731488216.311579&cid=CJ6C33JS0)
(We are actually not going to support the fonts in question anyway as none of them have a direct URL download, so it's kinda moot, but I just want to make sure we don't paint ourselves into a corner in our use of FLO)
[4:16](https://sil-lt.slack.com/archives/CJ6C33JS0/p1731489412342579?thread_ts=1731488216.311579&cid=CJ6C33JS0)
Wakor has fontRevision of 0x401CB which also doesn't really help!
[4:17](https://sil-lt.slack.com/archives/CJ6C33JS0/p1731489439489459?thread_ts=1731488216.311579&cid=CJ6C33JS0)
I am keen to avoid downloading hundreds of fonts in each build of the keyboards repository, so I want to cache, but if the version metadata is not reliable, that makes it tricky


Victor Gaultney
  9 minutes ago
head.fontRevision is not carefully handled by font providers and can be downriught odd


Marc Durdin
  [8 minutes ago](https://sil-lt.slack.com/archives/CJ6C33JS0/p1731489499403649?thread_ts=1731488216.311579&cid=CJ6C33JS0)
So name.Version is a safer bet, /^Version (\d+\.\d+)/i?


Victor Gaultney
  8 minutes ago
That too is not handled consistently - it’s just a string whose format cannot be guaranteed
[4:19](https://sil-lt.slack.com/archives/CJ6C33JS0/p1731489571841819?thread_ts=1731488216.311579&cid=CJ6C33JS0)
There is no ‘safe’ bet here


Marc Durdin
  [5 minutes ago](https://sil-lt.slack.com/archives/CJ6C33JS0/p1731489702880539?thread_ts=1731488216.311579&cid=CJ6C33JS0)
I see. I might have to maintain a map of ttf:name.Version:FLO-version -- that would allow me to invalidate stale fonts.
The other half is knowing when a font is updated in FLO data. For that, we are proposing to have a reference such as flo:abyssinicasil@2.300 in our data files (we also support other references such as GitHub raw URLs)


Victor Gaultney
  [2 minutes ago](https://sil-lt.slack.com/archives/CJ6C33JS0/p1731489896112729?thread_ts=1731488216.311579&cid=CJ6C33JS0)
I think you could use the name string regex as a first step, and if the result is a #.### pattern use it. If it’s not then you can’t compare.


Marc Durdin
  [1 minute ago](https://sil-lt.slack.com/archives/CJ6C33JS0/p1731489939508819?thread_ts=1731488216.311579&cid=CJ6C33JS0)
Yep, that's my thinking. And perhaps disallow use of fonts via FLO which don't meet that req. I'm okay with that (fallback is they can download the font and embed it the legacy way, or hassle the font author to up their game)
[4:26](https://sil-lt.slack.com/archives/CJ6C33JS0/p1731489987707979?thread_ts=1731488216.311579&cid=CJ6C33JS0)
... then if the FLO version doesn't match, we'd flag a version update for the containing keyboard package. And that way, we can (semi-)automatically update packages which have updates to fonts (which will make David and Lorna very happy).
Hard part is what to do when we do a build: we may not have all the fonts in cache, which could cause the build to fall over if a version is updated. I need to think this through
Stability is really important for us as we have to check over 1,000 packages in repo which may need updates when a font is updated, plus hundreds of keyboard authors who would also be impacted.
Reproducible keyboard package builds is an essential feature. Versioning of fonts is tricky in this respect (edited)

**Comment by mcdurdin** [comment]
Another capture

marc
  [Monday at 10:28 AM](https://keymanapp.slack.com/archives/C6Q9WS09G/p1731295710476959)
I have a remote referencing conundrum. kmc-copy and kmc-package are both now able to reference github URIs for content, but they have different patterns for referencing:
kmc-copy has: github:owner/repo:[branch:]path/to/file
kmc-package has: github:owner/repo/raw/git-hash/path/to/file
The purposes are slightly different. kmc-copy is a once off action, so we allow branch references. kmc-package  is for data in a .kps file, so should be referencing a permanent URI (hence use of the commit hash).
Options:
We could use a valid https url and match valid patterns, e.g. https://github.com/owner/repo/tree/branch/path/to/file  and https://github.com/owner/repo/raw/git-hash/path/to/file , which makes copypasta easier but may make it less obvious why we don't allow certain patterns.
We could use kmc-copy pattern for kmc-package : github:owner/repo:git-hash:path/to/file
We could use kmc-package pattern for kmc-copy : github:owner/repo/branch/path/to/file
Leave them divergent (ugh)
What do y'all think?
4 replies


ross
  [Monday at 1:47 PM](https://keymanapp.slack.com/archives/C6Q9WS09G/p1731307637151529?thread_ts=1731295710.476959&cid=C6Q9WS09G)
I think I like option 2.  Iike the / used for directory  : for different "named" identifiers.   Although I realise owner/repo still has a slash
Saved for later • Due 1 hour ago


eberhard
  [Today at 2:58 PM](https://keymanapp.slack.com/archives/C6Q9WS09G/p1731484683032619?thread_ts=1731295710.476959&cid=C6Q9WS09G)
I agree that option 1 is probably not the best since it suggests that you can use any URL.
Option 3 allows the user to make fewer changes to the copy/pasted URL, so I'd go with that.


marc
  [1 hour ago](https://keymanapp.slack.com/archives/C6Q9WS09G/p1731487402721689?thread_ts=1731295710.476959&cid=C6Q9WS09G)
Opt 3 should probably be: github:<owner>/<repo>/raw/<hash>/<path/to/file> for kmc-package, for consistency (and so we can allow other terms in place of 'raw' in future. This allows a right-click copy URL and just a very minor edit, for kmc-package.
For kmc-copy, I wonder if allowing a paste from GH is actually fine? Which then makes me wonder if a https URL is actually better for kmc-package as well, if we have clear messaging and UI around it. It means users can just open the URL to see what they get, as well.
So I am coming around to option 1.
Which now means we have 3 different answers to my question! Oops.
[3:44](https://keymanapp.slack.com/archives/C6Q9WS09G/p1731487447930879?thread_ts=1731295710.476959&cid=C6Q9WS09G)
The compiler can provide error messages with precise URL formatting. We can even translate [githubusercontent.com](http://githubusercontent.com/) urls to the original GH url in the IDE

**Comment by mcdurdin** [comment]
marc
  [Monday at 10:44 AM](https://keymanapp.slack.com/archives/C6Q9WS09G/p1731296653196259)
Another remote conundrum. We want to be able to support [fonts.languagetechnology.org](http://fonts.languagetechnology.org/) (FLO) for sourcing fonts for keyboard packages. However, we need to have stability, and FLO font versions and references can change without warning. Two issues:
Reproducible build requirement means we want a static font version -- how can we guarantee this?
We need to bump keyboard version when font version changes, so users get the new font. How can we learn of version changes to fonts? (And can we automate?)
4 replies


davidrowe
  [Monday at 1:31 PM](https://keymanapp.slack.com/archives/C6Q9WS09G/p1731306719472029?thread_ts=1731296653.196259&cid=C6Q9WS09G)
Would the commit hash scheme work to get a fixed reference?


marc
  [Monday at 1:32 PM](https://keymanapp.slack.com/archives/C6Q9WS09G/p1731306756212369?thread_ts=1731296653.196259&cid=C6Q9WS09G)
yes, commit hash is guaranteed fixed reference
Saved for later • Due 1 hour ago


eberhard
  [Today at 3:02 PM](https://keymanapp.slack.com/archives/C6Q9WS09G/p1731484944474909?thread_ts=1731296653.196259&cid=C6Q9WS09G)
Looking at the source of FLO (https://github.com/silnrsi/fonts) I see that the fonts seem to have a fontmanifest.json file that contains a version number, so we could query that repo for changes to fonts.

silnrsi/fonts
Collection of webfonts for internal use
Stars
5
Language
HTML
Added by [GitHub](https://keymanapp.slack.com/services/B01UP7G365R)


marc
  [43 minutes ago](https://keymanapp.slack.com/archives/C6Q9WS09G/p1731487601151899?thread_ts=1731296653.196259&cid=C6Q9WS09G)
Yeah, the FLO data includes a version number, e.g. "version": "1.490", for this font below.
"nokyung": {
    "defaults": {
        "ttf": "Nokyung-Regular.ttf",
        "woff": "Nokyung-Regular.woff"
    },
    "distributable": true,
    "family": "Nokyung",
    "familyid": "nokyung",
    "files": {
        "Nokyung-Bold.ttf": { "axes": { "ital": 0, "wght": 700.0 }, "flourl": "https://fonts.languagetechnology.org/fonts/sil/nokyung/Nokyung-Bold.ttf", "packagepath": "Nokyung-Bold.ttf", "url": "https://github.com/silnrsi/fonts/raw/main/fonts/sil/nokyung/Nokyung-Bold.ttf", "zippath": "Nokyung-1.490-dev-e7e323M/Nokyung-Bold.ttf" },
        "Nokyung-Bold.woff": { "axes": { "ital": 0, "wght": 700.0 }, "flourl": "https://fonts.languagetechnology.org/fonts/sil/nokyung/web/Nokyung-Bold.woff", "packagepath": "web/Nokyung-Bold.woff", "url": "https://github.com/silnrsi/fonts/raw/main/fonts/sil/nokyung/web/Nokyung-Bold.woff", "zippath": "Nokyung-1.490-dev-e7e323M/web/Nokyung-Bold.woff" },
        "Nokyung-Regular.ttf": { "axes": { "ital": 0, "wght": 400.0 }, "flourl": "https://fonts.languagetechnology.org/fonts/sil/nokyung/Nokyung-Regular.ttf", "packagepath": "Nokyung-Regular.ttf", "url": "https://github.com/silnrsi/fonts/raw/main/fonts/sil/nokyung/Nokyung-Regular.ttf", "zippath": "Nokyung-1.490-dev-e7e323M/Nokyung-Regular.ttf" },
        "Nokyung-Regular.woff": { "axes": { "ital": 0, "wght": 400.0 }, "flourl": "https://fonts.languagetechnology.org/fonts/sil/nokyung/web/Nokyung-Regular.woff", "packagepath": "web/Nokyung-Regular.woff", "url": "https://github.com/silnrsi/fonts/raw/main/fonts/sil/nokyung/web/Nokyung-Regular.woff", "zippath": "Nokyung-1.490-dev-e7e323M/web/Nokyung-Regular.woff" }
    },
    "license": "OFL",
    "packageurl": "https://software.sil.org/downloads/r/nokyung/Nokyung-1.490-dev-e7e323M.zip",
    "siteurl": "https://software.sil.org/Nokyung/",
    "source": "SIL",
    "status": "current",
    "version": "1.490",
    "ziproot": "Nokyung-1.490-dev-e7e323M"
},
Then the only pattern we need is the version, so do we have flo:nokyung@1.490 or flo:nokyung/1.490 or something else?

**Comment by mcdurdin** [comment]
My current thinking on flo: references. Perhaps FLO needs to be a tool in Developer which resolves to a GH stable commit reference, as fonts with a url are all GitHub references in FLO. 

If we always use GH, then it will be exceptionally rare to have a stable uri disappear -- only if repos are removed or nasty actions like force pushes are done -- in which case we probably want to deal with it anyway.

Additional feature: we can keep the original flo references as well, and use that to provide a hint message to flag when font updates are available, but do not attempt to use it directly.

File: developer/docs/help/reference/file-types/kps.md [hunk 1]
```diff
@@ -17,11 +17,156 @@ Details:
...
+We will add a `Source` element to .kps `Package.Files.File`. We would support
+several sources:
+* local - local file system
+* flo - fonts.languagetechnology.org
+* github
+* noto - in the future
+
+## local: `<Source>` element is omitted
+
+This is the same as previous versions. `<Name>` references a file on the local
```
Comment (darcywong00) [inline developer/docs/help/reference/file-types/kps.md:54]: What does  "... same as previous versions." refer to?

File: developer/docs/help/reference/file-types/kps.md [hunk 2]
```diff
@@ -17,11 +17,156 @@ Details:
...
+```
+
+We will add a `Source` element to .kps `Package.Files.File`. We would support
+several sources:
+* local - local file system
+* flo - fonts.languagetechnology.org
+* github
+* noto - in the future
+
+## local: `<Source>` element is omitted
```
Comment (darcywong00) [inline developer/docs/help/reference/file-types/kps.md:52]: This took me a few re-reads of the part about "`<Source>` element is omitted."
That phrase doesn't appear in the other source headings (flo, github, noto).

Is this conveying for local sources, the `<Source>` element will **always** be omitted? 
vs other sources the `<Source>` element can be omitted?

File: developer/docs/help/reference/file-types/kps.md (hunk 1 above)
Comment (mcdurdin) [reply developer/docs/help/reference/file-types/kps.md:54]: Hmm, this means v7, v12 of the format. I can update.

File: developer/docs/help/reference/file-types/kps.md [hunk 3]
```diff
@@ -17,11 +17,156 @@ Details:
...
+
+This is the same as previous versions. `<Name>` references a file on the local
+filesystem. Note that `<Source>` is optional for other sources as well; the
+absence of the element does not mean local -- that is determined by pattern
+match.
```
Comment (mcdurdin) [inline developer/docs/help/reference/file-types/kps.md:57]:
```suggestion
If `<Name>` references a file on the local filesystem, then the
`<Source>` element must be omitted. This matches versions 7.0 and 12.0 
of the .kps format for backward compatibility.

Note that `<Source>` is optional for other sources as well; the
absence of the element does not mean local -- that is determined by pattern
match.
```

File: developer/docs/help/reference/file-types/kps.md [hunk 4]
```diff
@@ -17,11 +17,156 @@ Details:
...
+```
+<Name>https://github.com/{owner}/{repo}/raw/{commit}/[path/]{filename}</Name>
+```
+
+We will add a `Source` element to .kps `Package.Files.File`. We would support
+several sources:
+* local - local file system
+* flo - fonts.languagetechnology.org
+* github
+* noto - in the future
```
Comment (mcdurdin) [inline developer/docs/help/reference/file-types/kps.md:50]:
```suggestion
There may a separate `<Source>` element as well, representing the following possible original sources for a file:
* Local file system -- omit the `<Source>` element
* fonts.languagetechnology.org -- `<Source>` should have the pattern `flo:...` 
* `https://github.com/...` -- `<Source>` should have a URL to a file stored in GitHub
* `noto:...` - to be supported in the future

The `<Source>` element is always optional; if it is not present, then kmc-package will not check for updated versions of the source file.
```

File: developer/src/kmc-package/test/fixtures/invalid/warn_font_in_flo_does_not_have_default_ttf.kps [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+  <Files>
+    <File>
+      <Name>basic.kmx</Name>
+      <Description>Keyboard Basic</Description>
+      <CopyLocation>0</CopyLocation>
+      <FileType>.kmx</FileType>
+    </File>
+    <!-- warn_font_in_flo_does_not_have_default_ttf: this uses a mismatching font, but
+         we won't get that far. This handles the scenario where FLO previously lists a font
+         but no ttf is available -->
```
Comment (darcywong00) [inline developer/src/kmc-package/test/fixtures/invalid/warn_font_in_flo_does_not_have_default_ttf.kps:22]: In the future when FLO makes updates, are there scenarios that could that impact some text fixture assumptions? 

For example, a listed font that previously doesn't have ttf changes to include one?

**Review by darcywong00 — APPROVED** [review]
A beefy PR!

File: developer/src/kmc-package/test/fixtures/invalid/keyboardcontentfilenotfound.kps [hunk 6]
```diff
@@ -24,84 +24,6 @@
...
-      <CopyLocation>0</CopyLocation>
-      <FileType>.ttf</FileType>
-    </File>
-    <File>
-      <Name>splash.gif</Name>
-      <Description>File splash.gif</Description>
-      <CopyLocation>0</CopyLocation>
-      <FileType>.gif</FileType>
-    </File>
   </Files>
```
[the remark this answers is not in the document]
Comment (mcdurdin) [reply developer/src/kmc-package/test/fixtures/invalid/keyboardcontentfilenotfound.kps:27]: `khmer_angkor.kmx` is missing, have added a comment

File: developer/src/kmc-package/test/fixtures/invalid/warn_font_in_flo_does_not_have_default_ttf.kps (hunk 5 above)
Comment (mcdurdin) [reply developer/src/kmc-package/test/fixtures/invalid/warn_font_in_flo_does_not_have_default_ttf.kps:22]: The fixtures are all stored offline to avoid this arising (and have performant and consistent tests), see https://github.com/keymanapp/keyman/blob/c6baddad1237e09a6cdc14ccc64dd6d63253d63a/developer/src/kmc-package/test/fixtures/online/fonts.languagetechnology.org/%23families.json for example

**Comment by mcdurdin** [comment]
@ermshiperete and @srl295 would love your feedback on this before merge, particularly considering systemic issues (docs are included in the PR; most of the files are fixtures for tests). See also the TODO list in #11236 for some things I am already thinking about.

File: developer/docs/help/reference/file-types/kps.md [hunk 7]
```diff
@@ -17,11 +17,156 @@ Details:
...
+The `Name` element may be either a local filename, or a GitHub raw permanent
+URL, conforming to the pattern:
+
+```
+<Name>https://github.com/{owner}/{repo}/raw/{commit}/[path/]{filename}</Name>
+```
+
+We will add a `Source` element to .kps `Package.Files.File`. We would support
+several sources:
+* local - local file system
```
Comment (srl295) [inline developer/docs/help/reference/file-types/kps.md:47]:
```suggestion
We will add a `Source` element to .kps `Package.Files.File`. We would support
several sources:
```

…

I don't think local should even be in the list.  It's a bit jarring, as if to have:

> Oil:
> - Heating: fuel oil
> - Frying: vegetable oil
> - Cars: motor oil
> - Salad: olive oil
> - Tortillas: Don't use oil when making tortillas

File: developer/src/kmc-package/src/compiler/get-file-data.ts [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+    }
+  } catch(e) {
+    callbacks.reportMessage(PackageCompilerMessages.Error_CouldNotRetrieveStableUriFromGitHub({url, e}));
+    throw e;
+  }
+
+  return `https://github.com/${matches.groups.name}/${matches.groups.repo}/raw/${commit.sha}/${matches.groups.path}`;
+}
+
+async function getFileDataFromGitHub(callbacks: CompilerCallbacks, inputFilename: string, matches: RegExpExecArray): Promise<KmpCompilerFileDataResult> {
```
Comment (ermshiperete) [inline developer/src/kmc-package/src/compiler/get-file-data.ts:157]: To keep the order the same as with the previous functions where we define the outer function first and then the function that gets called by the outer function, you could move this function after `getFileDataFromRemote`. Would make it easier to read when reading the whole file.

File: developer/src/kmc-package/src/compiler/kmp-compiler.ts [hunk 9]
```diff
@@ -468,49 +478,25 @@ export class KmpCompiler implements KeymanCompiler {
 
     const hasKmpInf = !!data.files.find(file => file.name == KMP_INF_FILENAME);
 
-    let failed = false;
-    data.files.forEach((value) => {
+    for(const value of data.files) {
```
Comment (ermshiperete) [inline developer/src/kmc-package/src/compiler/kmp-compiler.ts:481]: Out of curiosity: why did you change `data.files.forEach` to a `for` loop?

File: developer/src/kmc-package/src/compiler/package-compiler-messages.ts [hunk 10]
```diff
@@ -146,5 +148,126 @@ export class PackageCompilerMessages {
...
+  static ERROR_SourceCannotBeSetForLocalFiles = SevError | 0x0030;
+  static Error_SourceCannotBeSetForLocalFiles = (o:{filename: string, sourceFilename: string}) => m(
+    this.ERROR_SourceCannotBeSetForLocalFiles,
+    `The '<Source>' element cannot be set for local file ${o.filename}`,
+  );
+
+  static HINT_RemoteReferencesShouldBeVersion18Plus = SevHint | 0x0031;
+  static Hint_RemoteReferencesShouldBeVersion18Plus = (o:{filename: string, kpsVersion: string}) => m(
+    this.HINT_RemoteReferencesShouldBeVersion18Plus,
+    `The source package includes a reference to URL '${def(o.filename)}', and the package source version is '${def(o.kpsVersion)}'; the package source version should be at least '18.0'`,
```
Comment (ermshiperete) [inline developer/src/kmc-package/src/compiler/package-compiler-messages.ts:270]: Maybe?
```suggestion
    `The source package includes a reference to URL '${def(o.filename)}' but the package source version is '${def(o.kpsVersion)}'; the package source version should be at least '18.0'`,
```

File: developer/src/kmc-package/src/compiler/package-version-validator.ts [hunk 11]
```diff
@@ -4,9 +4,13 @@ import { KeyboardMetadataCollection } from './package-metadata-collector.js';
 import { KpsFile, CompilerCallbacks } from '@keymanapp/developer-utils';
 
 export const DEFAULT_KEYBOARD_VERSION = '1.0';
-export const MIN_LM_FILEVERSION_KMP_JSON = '12.0';
+
+/** minimum FileVersion for keyboard packages */
```
Comment (ermshiperete) [inline developer/src/kmc-package/src/compiler/package-version-validator.ts:8]: :+1:

File: developer/src/kmc-package/test/package-compiler.tests.ts [hunk 12]
```diff
@@ -267,21 +252,111 @@ describe('KmpCompiler', function () {
...
+
+    // https://github.com/silnrsi/fonts/raw/b88c7af5d16681bd137156929ff8baec82526560/fonts/sil/alkalami/Alkalami-Regular.ttf
+    const source = 'https://github.com/silnrsi/fonts/raw/b88c7af5d16681bd137156929ff8baec82526560/fonts/sil/alkalami/Alkalami-Regular.ttf';
+    const matches = getFileDataEndpoints.GITHUB_STABLE_SOURCE.exec(source);
+    assert.isNotNull(matches);
+    assert.equal(matches.groups.name, 'silnrsi');
+    assert.equal(matches.groups.repo, 'fonts');
+    assert.equal(matches.groups.hash, 'b88c7af5d16681bd137156929ff8baec82526560');
+    assert.equal(matches.groups.path, 'fonts/sil/alkalami/Alkalami-Regular.ttf');
+    const res = await getFileDataEndpoints.getFileDataFromGitHub(callbacks, '', matches);
```
Comment (ermshiperete) [inline developer/src/kmc-package/test/package-compiler.tests.ts:284]: Does this get the file from GitHub or is this mocked somehow?

**Comment by mcdurdin** [comment]
I think I have addressed all the comments in the reviews, thanks all!
